---
title: "CNprep"
author: "Alex Krasnitz, Guoli Sun"
date: "`r Sys.Date()`"
bibliography: CNprep_biblio.bibtex
output: 
    BiocStyle::html_document:
        toc: true
vignette: >
  %\VignetteIndexEntry{CNprep}
  %\VignettePackage{CNprep}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---  

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<br>
**Package**: `r packageDescription("CNprep")[["Package"]]`<br />
**Authors**: `r packageDescription("CNprep")[["Author"]]`<br />
**Version**: `r packageDescription("CNprep")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("CNprep")[["License"]]`<br />

# License

The *CNprep* package and the underlying *CNprep* code 
are distributed under the 
[GPL-2 License](https://www.gnu.org/licenses/old-licenses/gpl-2.0.txt). You 
are free to use and redistribute this software. 

# Citing

If you use this package for a publication, we would ask you to cite the 
following:

> Alex Krasnitz and Guoli Sun (2014). CNprep: Pre-process DNA Copy Number (CN) Data for Detection of CN Events. R package version 2.0. https://CRAN.R-project.org/package=CNprep


# Introduction

Copy number variations (CNVs) are associated with a wide spectrum of 
pathological conditions and complex traits, such 
as Tourette Disorder [@Wang2018] and especially 
cancer [@Pelttari2018; @Franch-Exposito2018a]. Furthermore, CNVs can 
potentially be used as biomarker for cancer 
diagnosis [@Pan2018].

One commonly used technology to detect CNVs is the Comparative Genomic 
Hybridization array (array CGH). From those arrays, software such as 
*DNACopy* [@Seshan] can identify regions of gained and lost copy number.
However, because of the low resolution of the array, detection of short 
CNVs is limited [@Duan].
The availability and low cost of Next Generation Sequencing (NGS)
have enabled precise detection of CNVs for the entire genome with 
the exclusion of few hard to map regions. This leaded to the 
development of software adapted to 
whole-genome and whole-exome sequencing, such as *CNVkit* [@Talevich] and 
*exomecopy* [@Love]. 

The *CNprep* package evaluates DNA copy number data, using both their initial 
form (copy number as a noisy function of genomic position) and their 
approximation by a piecewise-constant function (segmentation), for the purpose 
of identifying genomic regions where the copy number differs from the norm.


# Installing and loading *CNprep* package

The *CNprep* package can be installed from github website using 
*devtools* package:

```{r install, echo=TRUE, message=FALSE, warning=FALSE, eval=FALSE}
library(devtools)
devtools::install_github("belleau/CNprep")
```

As with any R package, the `r Rpackage("CNprep")` package has to be loaded 
with the following command:

```{r loadingPackage, warning=FALSE, message=FALSE}
library(CNprep)
```


# Processing copy number 


# Create and apply a mask to copy number events


## Make copy number mask

The *makeCNPmask()* function takes as an input a table of copy number events 
with genomic positions. First, interval regions where 
the event count is above the lower threshold are identified. Among those 
retained regions, the function only returns the interval regions that
contain at least one sub-interval with the count above 
the upper threshold.

As an example, a mask in created from copy number events detected in 10 
samples. In Figure \@ref(fig:demoMask), the first section represent
copy number events detected on 10 samples for a small region of chromosome 1.
The middle section shows the frequency of the events with the lower and upper
thresholds selected for this analysis. The last section displays the result of
the #makeCNPmask* function for this dataset when the lower threshold is set to
0.25 and upper threshold to 0.45.

```{r demoMask, echo=FALSE, message=FALSE, warning=FALSE, fig.align='center', fig.cap="Example of makeCNPmask() function. Copy number events detected in 10 samples in top section. Frequency of the events with the selected lower and upper thresholds in the middle section. Result of the makeCNPmask() function for those thresholds in the bottom section."}
library(GenomicRanges)
library(gtrellis)

dataTest <- data.frame(chrom=rep("chr1", 10), 
                start=c(12742927, 13071912, 12862827, 12708592, 13271912, 
                        13071612, 12838592, 13091712, 12842827, 12742827), 
                end=c(13039275, 13439275, 12939275, 13039175, 13551912, 
                        13191012, 12930000, 13542827, 13456522, 13706522), 
                value=seq(from=0.05, to=0.95, by=.1))

gr <- GRanges(seqnames = Rle(c("chr1"), c(10)),
    ranges = IRanges(start = dataTest$start, end = dataTest$end, 
                        names = head(letters, 10)),
    strand = Rle(strand(c("*")), c(10)))

covN <- as.data.frame(GRanges(coverage(gr)))
covN <- covN[-1,]
covN2 <- covN
covN2$start <- covN2$end
covN$end <-covN$start
covNT <- rbind(covN, covN2)
covNT <- covNT[order(covNT$start),]
covNT$freq <- covNT$score/10

#library(ComplexHeatmap)
#lgd = Legend(at = c("lower", "upper"), title = "Threshold", type = "points", legend_gp = gpar(col = 2:3))


minThres <- data.frame(chrom=rep("chr1", 1), start=c(12708592, 13706522), 
                        end=c(12708592, 13706522))

gtrellis_layout(data=dataTest, xpadding = c(0.015, 0.015), n_track = 3, 
                track_ylim = c(0, 1, 0, 0.70, 0, 1), add_name_track = TRUE,
                track_height = unit.c(unit(2, "null"), unit(4, "null"), 
                                        unit(1, "null")),
                axis_label_fontsize=13, lab_fontsize=15,
                track_axis = c(FALSE, TRUE, FALSE),  
                track_ylab = c("Copy Number Event", "Event Frequency", "Mask"))

## Add segments representing the copy number events
add_segments_track(dataTest, value=dataTest$value, gp=gpar(col="darkorange3", 
                                                            lwd=4))

## Adding the frequency (nb events/nb samples)
add_lines_track(covNT, value=covNT$freq, gp=gpar(lwd=3, col="darkblue"))

## Adding lower threshold
add_lines_track(minThres, value=0.25, gp=gpar(col="red3", lty=4, lwd=2), 
                    track=current_track())

## Adding upper threshold
add_lines_track(minThres, value=0.45, gp=gpar(col="red3", lty=3, lwd=2), 
                    track=current_track())

dataTest2 <- dataTest
dataTest2$chrom <- rep(1, 10)

mask <- makeCNPmask(imat=dataTest2, chromCol="chrom",
    startCol="start", endCol="end", nProf=10,
    uthresh=0.45, dthresh=0.25)
mask <- as.data.frame(mask)
mask$chrom <- rep("chr1", nrow(mask))

## Add segments representing the copy number events
add_segments_track(mask, value=0.5, gp=gpar(col="darkorange4", lwd=6))
```


Using a upper threshold of 0.55 on the same copy number events detected on 10 
samples results in only one genomic region in the mask, as shown in 
Figure \@ref(fig:demoMask2). This region is the only one to respect both 
lower and upper thresholds. 

```{r demoMask2, echo=FALSE, message=FALSE, warning=FALSE, fig.align='center', fig.cap="Example of makeCNPmask() function with stringent parameters. Copy number events detected in 10 samples in top section. Frequency of the events with the selected lower and upper thresholds in the middle section. Result of the makeCNPmask() function for those thresholds in the bottom section."}
library(GenomicRanges)
library(gtrellis)
# 
# dataTest <- data.frame(chrom=rep("chr1", 10), 
#                 start=c(12742927, 13071912, 12862827, 12708592, 13271912, 
#                     13071612, 12838592, 13091712, 12842827, 12742827), 
#                 end=c(13039275, 13439275, 12939275, 13039175, 13551912, 
#                     13191012, 12930000, 13542827, 13456522, 13706522), 
#                 value=seq(from=0.05, to = 0.95, by = .1))

gr <- GRanges(seqnames = Rle(c("chr1"), c(10)),
        ranges=IRanges(start=c(12742927, 13071912, 12862827, 12708592, 
                13271912, 13071612, 12838592, 13091712, 12842827, 12742827), 
            end=c(13039275, 13439275, 12939275, 13039175, 13551912, 
                13191012, 12930000, 13542827, 13456522, 13706522),
            names = head(letters, 10)), strand=Rle(strand(c("*")), c(10)))

covN <- as.data.frame(GRanges(coverage(gr)))[-1,]
covN2 <- covN
covN2$start <- covN2$end
covN$end <-covN$start
covNT <- rbind(covN, covN2)
covNT <- covNT[order(covNT$start),]
covNT$freq <- covNT$score/10

#library(ComplexHeatmap)
#lgd = Legend(at = c("lower", "upper"), title = "Threshold", type = "points", legend_gp = gpar(col = 2:3))


thresholdInfo <- data.frame(chrom=rep("chr1", 1), start=c(12708592, 13706522), 
                            end=c(12708592, 13706522))

## Create trellis graph
gtrellis_layout(data=gr, xpadding = c(0.015, 0.015), n_track=3, 
                track_ylim = c(0, 1, 0, 0.70, 0, 1), add_name_track=TRUE,
                track_height = unit.c(unit(2, "null"), unit(4, "null"), 
                                        unit(1, "null")),
                track_axis = c(FALSE, TRUE, FALSE),  
                axis_label_fontsize=13, lab_fontsize=15,
                track_ylab = c("Copy Number Event", "Event Frequency", "Mask"))

## Add segments representing the copy number events
add_segments_track(gr, value=dataTest$value, gp=gpar(col="darkorange3", 
                                                            lwd=4))

## Adding the frequency (nb events/nb samples)
add_lines_track(covNT, value=covNT$freq, gp=gpar(lwd=3, col="darkblue"))

## Adding lower threshold
add_lines_track(thresholdInfo, value=0.25, gp=gpar(col="red3", lty=4, lwd=2), 
                    track=current_track())

## Adding upper threshold
add_lines_track(thresholdInfo, value=0.55, gp=gpar(col="red3", lty=3, lwd=2), 
                    track=current_track())

dataTest2 <- as.data.frame(gr)
dataTest2$chrom <- rep(1, 10)

mask <- makeCNPmask(imat=dataTest2, chromCol="chrom",
    startCol="start", endCol="end", nProf=10,
    uthresh=0.55, dthresh=0.25)
mask <- as.data.frame(mask)
mask$chrom <- rep("chr1", nrow(mask))

## Add segments representing the copy number events
add_segments_track(mask, value=0.5, gp=gpar(col="darkorange4", lwd=6))
```



```{r makeMask, echo=TRUE, warning=FALSE, collapse=TRUE}
## Load example containing the copy number events for 1203 samples
data(cnpexample)

## Retaining only the amplification events
ampCNP <- subset(cnpexample, copy.num == "amp")

## The amplification table has 4 columns:
## copy.num identifies the type of event
## chrom is the number identifying the chromosome
## chrom.start is the starting position of the event
## Chrom.end is the ending psoition of the event
head(ampCNP, n=4)

## Create a mask using the amplification events.
## The name of the columns for the chromosome 
## identification (chromCol parameter), the
## starting and ending positions (startCol and endCol parameters) must 
## be given.
## The number of samples on which the events have been detected  
## is needed (nProf parameter).
ampCNPmask <- makeCNPmask(imat=ampCNP, chromCol="chrom",
    startCol="chrom.start", endCol="chrom.end", nProf=1203,
    uthresh=0.2, dthresh=0.01)

```



## Apply mask to a table of copy number events

The *applyCNPmask* function

# Session info

Here is the output of sessionInfo() on the system on which this document 
was compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References