---
title: "CNprep"
author: "Alex Krasnitz, Guoli Sun"
date: "`r Sys.Date()`"
bibliography: CNprep_biblio.bibtex
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{CNprep}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---  

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<br>
**Package**: `r packageDescription("CNprep")[["Package"]]`<br />
**Authors**: `r packageDescription("CNprep")[["Author"]]`<br />
**Version**: `r packageDescription("CNprep")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("CNprep")[["License"]]`<br />

## License

The *CNprep* package and the underlying *CNprep* code 
are distributed under the [GPL-2 License](https://www.gnu.org/licenses/old-licenses/gpl-2.0.txt). You are free to use and 
redistribute this software. 

## Citing


If you use this package for a publication, we would ask you to cite the 
following:

> Alex Krasnitz and Guoli Sun (2014). CNprep: Pre-process DNA Copy Number (CN) Data for Detection of CN Events. R package version 2.0. https://CRAN.R-project.org/package=CNprep


# Introduction

Copy number variation is a significant source of variation in the human genome.
Copy number variants (CNVs) has been associated with a number of diseases, such 
as Tourette Disorder [@Wang2018] and especially 
cancer [@Pelttari2018; @Franch-Exposito2018a]. Even 
more, CNVs can potentially to be used biomarker for cancer 
diagnosis [@Pan2018]. The spectrum of alterations, in somatic DNA copy number 
variation, is broad and includes both amplifications and deletions ranging 
from less than a single gene to an entire chromosome [@Kendall2011].

Estimation of DNA copy number genome-wide can be done using  high throughput 
sequencing methods. The DNA copy number are inferred according to
the hypothesis that the read depth in a region of the genome is roughly 
proportional to the DNA copy number [@Kendall2011].

The *CNprep* package evaluates DNA copy number data, using both their initial 
form (copy number as a noisy function of genomic position) and their 
approximation by a piecewise-constant function (segmentation), for the purpose 
of identifying genomic regions where the copy number differs from the norm.


# Session info

Here is the output of sessionInfo() on the system on which this document 
was compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References